Modules/_functoolsmodule.c

```diff
@@ -127,69 +130,144 @@ partial_dealloc(partialobject *pto)
 }
 
 static PyObject *
-partial_call(partialobject *pto, PyObject *args, PyObject *kw)
+partial_fastcall(partialobject *pto, PyObject **args, Py_ssize_t nargs,
+                 PyObject *kwargs)
 {
+    PyObject *small_stack[_PY_FASTCALL_SMALL_STACK];
     PyObject *ret;
-    PyObject *argappl, *kwappl;
-    PyObject **stack;
-    Py_ssize_t nargs;
+    PyObject *kwargs2;
+    PyObject **stack, **stack_buf = NULL;
+    Py_ssize_t nargs2, pto_nargs;
+
+    pto_nargs = PyTuple_GET_SIZE(pto->args);
+    if (pto_nargs == 0) {
+        stack = args;
+        nargs2 = nargs;
+    }
+    else if (nargs == 0) {
+        stack = &PyTuple_GET_ITEM(pto->args, 0);
+        nargs2 = pto_nargs;
+    }
+    else {
+        nargs2 = pto_nargs + nargs;
+        if (nargs2 <= (Py_ssize_t)Py_ARRAY_LENGTH(small_stack)) {
+            stack = small_stack;
+        }
+        else {
+            stack_buf = PyMem_Malloc(nargs2 * sizeof(PyObject *));
+            if (stack_buf == NULL) {
+                PyErr_NoMemory();
+                return NULL;
+            }
+            stack = stack_buf;
+        }
 
-    assert (PyCallable_Check(pto->fn));
-    assert (PyTuple_Check(pto->args));
-    assert (PyDict_Check(pto->kw));
+        /* use borrowed references */
+        memcpy(stack,
+               &PyTuple_GET_ITEM(pto->args, 0),
+               pto_nargs * sizeof(PyObject*));
+        memcpy(&stack[pto_nargs],
+               args,
+               nargs * sizeof(PyObject*));
+    }
+
+    if (PyDict_GET_SIZE(pto->kw) == 0) {
+        /* kwargs can be NULL */
+        kwargs2 = kwargs;
+        Py_XINCREF(kwargs2);
+    }
+    else {
+        /* bpo-27840, bpo-29318: dictionary of keyword parameters must be
+           copied, because a function using "**kwargs" can modify the
+           dictionary. */
+        kwargs2 = PyDict_Copy(pto->kw);
+        if (kwargs2 == NULL) {
+            PyMem_Free(stack_buf);
+            return NULL;
+        }
+
+        if (kwargs != NULL) {
+            if (PyDict_Merge(kwargs2, kwargs, 1) != 0) {
+                PyMem_Free(stack_buf);
+                Py_DECREF(kwargs2);
+                return NULL;
+            }
+        }
+    }
+
+    ret = _PyObject_FastCallDict(pto->fn, stack, nargs2, kwargs2);
+    PyMem_Free(stack_buf);
+    Py_XDECREF(kwargs2);
+    return ret;
+}
+
+static PyObject *
+partial_call_impl(partialobject *pto, PyObject *args, PyObject *kwargs)
+{
+    PyObject *ret, *args2, *kwargs2;
 
     if (PyTuple_GET_SIZE(pto->args) == 0) {
-        stack = &PyTuple_GET_ITEM(args, 0);
-        nargs = PyTuple_GET_SIZE(args);
-        argappl = NULL;
+        Py_INCREF(args);
```

Comment:
I tested: removing the two special case (size==0) has no impact on performance, so I removed it.